lib/Dialect/FIRRTL/FieldRef.cpp

```diff
@@ -0,0 +1,127 @@
+//===- FieldRef.cpp - FIRRTL Field Refs  ------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This header file defines FieldRef and helpers for them.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/FieldRef.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace firrtl;
+
+Operation *FieldRef::getDefiningOp() const {
+  if (auto *op = value.getDefiningOp())
+    return op;
+  return value.cast<BlockArgument>().getOwner()->getParentOp();
+}
+
+FieldRef::FieldRef(Value val) : value(val) {
```

Comment:
Should this functionality be in firrtl operations instead?  Only a few ops have non-trivial field-refs.  E.g. ops know how to return a fieldref for themselves?  wouldn't cover BlockArguments.

Reply:
I was thinking that we would only need to support SubIndex, SubField, and possibly SubAccess.  I didn't realize that SubAccess can be on the LHS side until earlier today, and I'm not sure how to deal with that yet.  I think this might be enough to make FieldRefs usable for LowerTypes (if we wanted).

It would be cool to implement this as an OpInterface, and maybe make this non-firrtl specific? E.g.
```
FieldRefResultOpInterface {
  getFieldRef(unsigned resultNo);
}
 ```
 For ops that don't implement it, it returns a field ref pointing to the result of the op. 

Similar idea: This has centralized handling of the names of results and ops. I think it might be good to have an OpInterface to handle things with names that was sharable with the RTL dialect.

Reply:
I moved `FieldRef` in to `circt/Support/` and moved the FIRRTL specific logic in to `FIRRTLDialect.h`, and we can figure out how to make this more generic later on.